feat(WIP): Try noports wish application

xavierchanth opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines in CONTRIBUTING.md

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**

- Created a TUI bubbletea application to be served over SSH
- Allows to run 3 commands:
  - Display a welcome message
  - Run `ip addr` / `ifconfig` (-f flag to use `ifconfig` instead of `ip addr`)
  - Run `nmap` to scan all ports for the given host passed in (-h flag)

Update (June 13th, 2024):
This was blocked by the C daemon, so we could statically link it into the go program.
This is now ready to move forward, but I (@XavierChanth) don't currently have the bandwidth to continue working on it.

**- How I did it**

**- How to verify it**

**- Description for the changelog**
feat: Try noports wish application
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

xavierchanth committed:
feat: initial trynoports prototype

xavierchanth committed:
chore: remove example comment

xavierchanth committed:
fix: esc yeets you out of the program

xavierchanth committed:
refactor: organize the try wish app code

xavierchanth committed:
chore: better styling

xavierchanth committed:
feat: get command runner working

xavierchanth committed:
Merge branch 'trunk' into try-noports-wish

xavierchanth committed:
feat: add viewport scrolling

xavierchanth committed:
chore: use ctrl+u / ctrl+d for vp scroll

xavierchanth committed:
chore: remove unused functions

xavierchanth committed:
Merge branch 'trunk' into try-noports-wish

xavierchanth committed:
feat: add welcome message

xavierchanth committed:
chore: cleanup the wish app

xavierchanth committed:
fix: log message

xavierchanth commented:
Forgot to finish the welcome message, I will also explain all this code in the files changed tab before EOD

xavierchanth committed:
chore: cleanup help casing and length

xavierchanth commented on packages/go/trynoports/.gitignore:
```diff
@@ -0,0 +1 @@
+.ssh/
```
App automatically generates a .ssh folder for keys, so ignore it

xavierchanth commented on packages/go/trynoports/app.go:
```diff
@@ -0,0 +1,130 @@
+)
+
+// Main application state, composed of 3 sub-models
+type appState struct {
+	program  *tea.Program
+	list     appList
+	frame    appFrame
+	viewport appViewport
+	width    int
+	height   int
+	focused  int
+}
```
Main application state, 3 sub-structs below

xavierchanth commented on packages/go/trynoports/app.go:
```diff
@@ -0,0 +1,130 @@
+package main
+
+// Model / Init part of Elm architecture for app
```
Elm architecture has 3 parts:

1. Model / Init (i.e. state)
2. Update (i.e. update the state based on an event)
3. View (i.e. presentation based on the state)

xavierchanth commented on packages/go/trynoports/app.go:
```diff
@@ -0,0 +1,130 @@
+		p := tea.NewProgram(m, opts...)
+		go func() {
+			// register a pointer to the program with itself, so we can send async messages back to the viewport later
+			p.Send(ProgramMsg{p})
```
Pass a pointer to the main program object into the program itself, so we can pass events asynchronously back to the program (such as rendering text from a command execution pipe)

xavierchanth commented on packages/go/trynoports/app.go:
```diff
@@ -0,0 +1,130 @@
+}
+type CmdDoneMsg struct{}
+
+func AppMiddleware() wish.Middleware {
```
Middleware that allows the bubbletea app to run on the wish ssh server

xavierchanth commented on packages/go/trynoports/app.go:
```diff
@@ -0,0 +1,130 @@
+		m := appState{
+			frame: appFrame{
+				width:  pty.Window.Width,
+				height: pty.Window.Height,
+				style:  frameStyle,
+			},
+			list: appList{
+				style: listStyle,
+				model: InitCommands(list.NewDefaultDelegate(), 0, appHeight-listStyle.GetVerticalFrameSize()),
+			},
+			viewport: appViewport{
+				style:   viewportStyle,
+				spinner: spin,
+			},
+		}
+		m.list.model.Title = title
+		m.list.model.AdditionalShortHelpKeys = func() []key.Binding {
+			return []key.Binding{
+				key.NewBinding(
+					key.WithKeys("enter"),
+					key.WithHelp("enter", "run command"),
+				),
+			}
+		}
+		m.list.model.AdditionalFullHelpKeys = func() []key.Binding {
+			return []key.Binding{
+				key.NewBinding(
+					key.WithKeys(tea.KeyCtrlD.String()),
+					key.WithHelp("ctrl+d", "down (viewport)"),
+				),
+				key.NewBinding(
+					key.WithKeys(tea.KeyCtrlU.String()),
+					key.WithHelp("ctrl+u", "up (viewport)"),
+				),
+				key.NewBinding(
+					key.WithKeys("enter"),
+					key.WithHelp("enter", "run command"),
+				),
+			}
+		}
```
Main state initialization, the viewport and list will have some additional state initialization done asynchronously later

xavierchanth commented on packages/go/trynoports/app_update.go:
```diff
@@ -0,0 +1,162 @@
+		m.program = msg.program
+	case CmdDoneMsg:
+		m.viewport.isRunning = false
+	case ViewportContentMsg:
+		m.viewport.content += msg.content
+		m.viewport.model.SetContent(m.viewport.content)
+		m.viewport.model.GotoBottom()
+		if useHighPerformanceRenderer {
+			viewport.Sync(m.viewport.model)
+		}
```
Asynchronous message for content to add to the viewport

xavierchanth commented on packages/go/trynoports/app_update.go:
```diff
@@ -0,0 +1,162 @@
+	switch msg := msg.(type) {
+	case ProgramMsg:
+		m.program = msg.program
+	case CmdDoneMsg:
+		m.viewport.isRunning = false
```
Allow the viewport to receive a new command when we receive the done message

xavierchanth commented on packages/go/trynoports/app_update.go:
```diff
@@ -0,0 +1,162 @@
+			viewport.Sync(m.viewport.model)
+		}
+
+		done := make(chan int, 1)
+		ch, err := m.list.model.SelectedItem().(RunnerWithDone).Run(done)
+		if err != nil {
+			close(done)
+			// Welcome command should never return an error, so it is safe to assume we have an appCommand in here
+			// Thus command & args are set
+			m.viewport.content = fmt.Sprintf("> %s %s\nOops! We messed up, can't run this command right now...", command, args)
+			m.viewport.model.SetContent(m.viewport.content)
+			m.viewport.model.GotoBottom()
+			m.viewport.isRunning = false
+			log.Error("Error running command: ", "error", err)
+			return
+		}
```
Run the command and immediately handle a failed startup

xavierchanth commented on packages/go/trynoports/app_update.go:
```diff
@@ -0,0 +1,162 @@
+			return
+		}
+
+		go func() {
+			for {
+				select {
+				case <-done:
+					close(ch)
+					close(done)
+					m.program.Send(CmdDoneMsg{})
+					return
+				case msg := <-ch:
+					m.program.Send(ViewportContentMsg{msg})
+				}
+			}
+		}()
```
asynchronously receive messages from the channel we opened, then send them back to bubbletea to display to the viewport, send CmdDoneMsg when the program is done

xavierchanth committed:
fix: handle viewport init correctly

xavierchanth committed:
chore: better formatting for small message size

xavierchanth commented on packages/go/trynoports/app_update.go:
```diff
@@ -0,0 +1,167 @@
+
+		// We have an initial window size now
+		// Now it is safe to initialize the viewport
+		if !m.viewport.isReady {
+			// We will resize this to the correct size during the ResizeComponents call
+			m.viewport.model = viewport.New(1, 1)
+			m.viewport.model.SetContent(welcomeMessageContent)
+			m.viewport.model.HighPerformanceRendering = useHighPerformanceRenderer
+			m.viewport.model.KeyMap.Down.SetEnabled(false)
+			m.viewport.model.KeyMap.Up.SetEnabled(false)
+			m.viewport.isReady = true
+		}
```
Rest of the viewport initialization, had to wait for the program to report an initial window size.

xavierchanth committed:
perf: run less stuff prior to viewport.isReady

xavierchanth commented on packages/go/trynoports/app_view.go:
```diff
@@ -0,0 +1,34 @@
+
+// View part of Elm architecture for app
+
+func (m appState) View() string {
+	return m.frame.style.Render(
+		lipgloss.JoinHorizontal(
+			lipgloss.Top,
+			m.RenderList(m.list.style),
+			m.RenderViewport(m.viewport.style),
+		),
+	)
+}
```
Render the list to the left, viewport to the right

xavierchanth committed:
fix: wrapping bug

xavierchanth commented on packages/go/trynoports/app_update.go:
```diff
@@ -0,0 +1,168 @@
+
+		// We have an initial window size now
+		// Now it is safe to initialize the viewport
+		if !m.viewport.isReady {
+			// We will resize this to the correct size during the ResizeComponents call
+			m.viewport.model = viewport.New(1, 1)
+			m.viewport.content = welcomeMessageContent
+			m.viewport.model.HighPerformanceRendering = useHighPerformanceRenderer
+			m.viewport.model.KeyMap.Down.SetEnabled(false)
+			m.viewport.model.KeyMap.Up.SetEnabled(false)
+			m.viewport.isReady = true
+			if useHighPerformanceRenderer {
+				cmds = append(cmds, viewport.Sync(m.viewport.model))
+			}
+		}
```
Initialization of the viewport, we will use the welcome message as the default content.

xavierchanth commented on packages/go/trynoports/app_update.go:
```diff
@@ -0,0 +1,168 @@
+		m.viewport.model.Width = viewportW
+		m.viewport.model.Height = viewportH
+
+		content := lipgloss.NewStyle().Width(viewportW).Render(m.viewport.content)
+		m.viewport.model.SetContent(content)
```
After resizing the viewport, normalize the content so that it wraps properly in the terminal

xavierchanth commented on packages/go/trynoports/main.go:
```diff
@@ -0,0 +1,85 @@
+	log.Info("Environment", "nmap host (-h)", *Flagh, "use ifconfig (-f)", *Flagf)
+
+	// Create the server object with appropriate middleware
+	s, err := wish.NewServer(
+		wish.WithAddress(net.JoinHostPort(host, port)),
+		wish.WithHostKeyPath(keyPath),
+		wish.WithMiddleware(
+			AppMiddleware(),
+			activeterm.Middleware(), // Bubble Tea apps usually require a PTY.
+			logging.Middleware(),
+		),
+	)
```
Setup the actual SSH server with the appropriate options / middlewares

cpswan committed:
Merge branch 'trunk' into try-noports-wish

cpswan reviewed (commented):
@XavierChanth how do I run this? I've already tried the obvious `go get && go run main.go` but that just gets me:

```
# command-line-arguments
./main.go:59:4: undefined: AppMiddleware
```

What's your thinking on tests? And how we're going to host this?

xavierchanth commented:
> @XavierChanth how do I run this? I've already tried the obvious `go get && go run main.go` but that just gets me:
> 
> ```
> # command-line-arguments
> ./main.go:59:4: undefined: AppMiddleware
> ```
> 
> What's your thinking on tests? And how we're going to host this?

I've written the main package across multiple files, so you have to run via `go run .`

I'm not sure what to do about tests yet, most of the functions should be pure enough to unit test, and I think I could emulate some end2end tests. For the time being, I was thinking I setup a trial container to test against, and give the team access for testing and feedback. Ideally we release this with the APKAM installer, so I'm about on track.

Also, when I write the tests I can refactor the other files so that they are different, isolated packages and that way running `go run main.go` will work as expected.

xavierchanth committed:
Merge branch 'trunk' into try-noports-wish

xavierchanth committed:
refactor: split app and command into their own modules

xavierchanth committed:
chore: more cleanup

xavierchanth committed:
chore: add warning when nmap is not set (make default localhost)

xavierchanth committed:
chore: cleanup go code

xavierchanth committed:
chore: ignore built binary

xavierchanth committed:
chore: checkin partial docker work

xavierchanth committed:
Merge branch 'trunk' into try-noports-wish

xavierchanth committed:
chore: remove docker entrypoint

xavierchanth committed:
feat: add internal sshnpd runner

xavierchanth committed:
chore: set trynoports binary to docker entrypoint

xavierchanth committed:
chore: ignore .atsign folder

xavierchanth committed:
build: MVP golang wish build

xavierchanth committed:
chore: we getting there...

for some reason srv is failing to create a file at startup... but what
file? the socket file descriptor maybe???

xavierchanth committed:
docs: explain how the docker build works

xavierchanth commented:
Started dockerizing, this file explains how the usage is intended: https://github.com/atsign-foundation/noports/blob/try-noports-wish/packages/go/trynoports/docker.md

I designed the go application to spawn a go routine which runs sshnpd as a child process. This works fine, but SRV seems to crash... I think it has trouble binding sockets (see logs below)

FYI. The only dart:io classes used in srv are `Process`, `Socket`, `InternetAddress`.


At the moment the application crashes with:

```
2024/05/03 23:16:01 INFO Sshnpd |
  Msg=
  │ r.result is a bool
  │ INFO|2024-05-03 23:16:01.373459| sshnpd |svr.result is true
  │  bool, Result: true, SigningMetadata: {HashingAlgo: sha256, SigningAlgo: rsa2048, SignatureTimestamp: 2024-05-03 23:16:01.370563Z}
  │ INFO|2024-05-03 23:16:01.373442| sshnpd |sv
2024/05/03 23:16:01 INFO Sshnpd |
  Msg=
  │ r.result is a bool
  │ INFO|2024-05-03 23:16:01.373459| sshnpd |svr.result is true

2024/05/03 23:16:01 INFO Sshnpd |
  Msg=
  │ INFO|2024-05-03 23:16:01.373799| sshnpd |Setting up ports for tunnel session using dart (dart) from: @xavierchanth session: d5ae6098-6937-4f74-a0c6-787b61fc8264

2024/05/03 23:16:01 INFO Sshnpd |
  Msg=
  │ INFO|2024-05-03 23:16:01.392005|SrvImplExec|SrvImplExec.run(): executing /usr/local/at/srv -h uplink.xavierchanth.xyz -p 43219 --local-port 23234 --local-host localhost --multi --rv-auth --rv-e2ee

2024/05/03 23:16:01 INFO Sshnpd |
  Msg=
  │ 1780|SrvImplExec|rv stderr | #0      _Directory.createSync (dart:io/directory_impl.dart:124)
  │ INFO|2024-05-03 23:16:01.621786|SrvImplExec|rv stderr | #1      _File.createSync (dart:io/file_impl.dart:303)
  │ INFO|2024-05-03 23:16:01.621791|SrvImplExec|rv stde
2024/05/03 23:16:01 INFO Sshnpd |
  Msg=
  │ 1780|SrvImplExec|rv stderr | #0      _Directory.createSync (dart:io/directory_impl.dart:124)
  │ INFO|2024-05-03 23:16:01.621786|SrvImplExec|rv stderr | #1      _File.createSync (dart:io/file_impl.dart:303)
  │ INFO|2024-05-03 23:16:01.621791|SrvImplExec|rv stde
2024/05/03 23:16:01 INFO Sshnpd |
  Msg=
  │ mplExec|rv stderr | #4      _delayEntrypointInvocation.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:295)
  │ INFO|2024-05-03 23:16:01.621807|SrvImplExec|rv stderr | #5      _RawReceivePort._handleMessage (dart:isolate-patch/isolate_patch.dart:18
2024/05/03 23:16:01 INFO Sshnpd |
  Msg=
  │ mplExec|rv stderr | #4      _delayEntrypointInvocation.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:295)
  │ INFO|2024-05-03 23:16:01.621807|SrvImplExec|rv stderr | #5      _RawReceivePort._handleMessage (dart:isolate-patch/isolate_patch.dart:18
2024/05/03 23:16:01 INFO Sshnpd |
  Msg=
  │ 4)

2024/05/03 23:16:04 INFO Sshnpd |
  Msg=
  │ SEVERE|2024-05-03 23:16:04.573590| sshnpd |startNpt failed with unexpected error : TimeoutException after 0:00:03.000000: Future not completed

^C2024/05/03 23:31:57 INFO Stopping SSH server

```

The stuff indented under `Msg=` are the logs from sshnpd, which are relogged through the go program's logger.

xavierchanth commented:
This should be progressable at this point, but we don't have as much demand for it compared to the amount of effort it would be to embed the C version of sshnpd into this program. Closing for now
